runtime/server/auth/jwt.go

```diff
@@ -0,0 +1,223 @@
+		if !publicKey.Valid() {
+			return nil, fmt.Errorf("invalid signing key in JWKS")
+		}
+		publicJWKS.Keys = append(publicJWKS.Keys, publicKey)
+	}
+	publicJSON, err := json.Marshal(publicJWKS)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Issuer{
+		issuerURL:  issuerURL,
+		signingKey: signingKey,
+		publicJWKS: publicJSON,
+	}, nil
+}
+
+// NewEphemeralIssuer creates an Issuer using a generated JWKS.
+// It is useful for development and testing, but should not be used in production.
+func NewEphemeralIssuer(issuerURL string) (*Issuer, error) {
+	// NOTE: JWKS generation based on: https://github.com/go-jose/go-jose/blob/v3/jose-util/generate.go
+
+	// Generate RSA private key
+	rsaKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create JWK
+	jwk := jose.JSONWebKey{
+		Key:       rsaKey,
+		Algorithm: string(jose.RS256),
+		Use:       "sig",
+	}
+
+	// Set key ID based on JWK thumbprint
+	thumb, err := jwk.Thumbprint(crypto.SHA256)
+	if err != nil {
+		return nil, err
+	}
+	jwk.KeyID = base64.URLEncoding.EncodeToString(thumb)
+
+	// Create JWKS JSON
+	jwks := jose.JSONWebKeySet{Keys: []jose.JSONWebKey{jwk}}
+	jwksJSON, err := json.Marshal(jwks)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewIssuer(issuerURL, jwk.KeyID, jwksJSON)
+}
+
+// TokenOptions provides options for Issuer.NewToken.
+type TokenOptions struct {
+	AudienceURL         string
+	Subject             string
+	TTL                 time.Duration
+	SystemPermissions   []Permission
+	InstancePermissions map[string][]Permission
+}
+
+// NewToken issues a new JWT based on the provided options.
+func (i *Issuer) NewToken(opts TokenOptions) (string, error) {
+	now := time.Now()
+	claims := &jwtClaims{
+		RegisteredClaims: jwt.RegisteredClaims{
+			ExpiresAt: jwt.NewNumericDate(now.Add(opts.TTL)),
+			IssuedAt:  jwt.NewNumericDate(now),
+			NotBefore: jwt.NewNumericDate(now),
+			Issuer:    i.issuerURL,
+			Subject:   opts.Subject,
+			Audience:  []string{opts.AudienceURL},
+		},
+		System:    opts.SystemPermissions,
+		Instances: opts.InstancePermissions,
+	}
+
+	token := jwt.NewWithClaims(jwt.GetSigningMethod(i.signingKey.Algorithm), claims)
+	token.Header["kid"] = i.signingKey.KeyID
+	res, err := token.SignedString(i.signingKey.Key)
+	if err != nil {
+		return "", err
+	}
+
+	return res, nil
+}
+
+// WellKnownHandleFunc serves the public keys of the Issuer's JWKS.
+// The Audience expects it to be mounted on {issuerURL}/.well-known/jwks.json.
+func (i *Issuer) WellKnownHandleFunc(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	_, _ = w.Write(i.publicJWKS)
+}
+
+// Audience represents a receiver of tokens from Issuer.
+// The Audience is used by the runtime to parse claims from a JWT.
+// It parses and validates tokens and resolves permissions.
+// It refreshes its JWKS in the background from {issuerURL}/.well-known/jwks.json.
+type Audience struct {
```

Comment:
should this be called `Validator` ?

Reply:
Is that what others call it? I'm open to changing it. I just mirrored the "issuer" and "audience" terminology from the JWT spec.

Reply:
ok makes sense, lets keep it as it is